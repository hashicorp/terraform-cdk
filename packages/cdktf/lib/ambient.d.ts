// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
declare module 'archiver';