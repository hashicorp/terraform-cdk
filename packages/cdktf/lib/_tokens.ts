// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Tokenization, DefaultTokenResolver, StringConcat } from "./tokens";
import { Construct } from "constructs";

const TOKEN_RESOLVER = new DefaultTokenResolver(new StringConcat());

// eslint-disable-next-line jsdoc/require-jsdoc
export function resolve<T>(scope: Construct, obj: T, preparing = false): T {
  return Tokenization.resolve(obj, {
    scope,
    preparing,
    resolver: TOKEN_RESOLVER,
  });
}
