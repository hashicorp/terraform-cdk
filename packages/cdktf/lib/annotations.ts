// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
// Originally from aws-cdk v2 because with constructs v10 addMessage was removed
// in favor of Annotations in the AWS CDK
// https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/annotations.ts
import { IConstruct } from "constructs";

export enum AnnotationMetadataEntryType {
  INFO = "@cdktf/info",
  WARN = "@cdktf/warn",
  ERROR = "@cdktf/error",
}

export const DISABLE_STACK_TRACE_IN_METADATA =
  "@cdktf/disable_stack_trace_in_metadata";

/**
 * Includes API for attaching annotations such as warning messages to constructs.
 */
export class Annotations {
  /**
   * Returns the annotations API for a construct scope.
   * @param scope The scope
   */
  public static of(scope: IConstruct) {
    return new Annotations(scope);
  }

  private readonly stackTraces: boolean;

  private constructor(private readonly scope: IConstruct) {
    const disableTrace =
      scope.node.tryGetContext(DISABLE_STACK_TRACE_IN_METADATA) || false;

    this.stackTraces = !disableTrace;
  }

  /**
   * Adds a warning metadata entry to this construct.
   *
   * The CLI will display the warning when an app is synthesized.
   * In a future release the CLI might introduce a --strict flag which
   * will then fail the synthesis if it encounters a warning.
   *
   * @param message The warning message.
   */
  public addWarning(message: string) {
    this.addMessage(AnnotationMetadataEntryType.WARN, message);
  }

  /**
   * Adds an info metadata entry to this construct.
   *
   * The CLI will display the info message when apps are synthesized.
   *
   * @param message The info message.
   */
  public addInfo(message: string): void {
    this.addMessage(AnnotationMetadataEntryType.INFO, message);
  }

  /**
   * Adds an { "error": <message> } metadata entry to this construct.
   * The toolkit will fail synthesis when errors are reported.
   * @param message The error message.
   */
  public addError(message: string) {
    this.addMessage(AnnotationMetadataEntryType.ERROR, message);
  }

  /**
   * Adds a message metadata entry to the construct node, to be displayed by the CDK CLI.
   * @param level The message level
   * @param message The message itself
   */
  private addMessage(level: AnnotationMetadataEntryType, message: string) {
    this.scope.node.addMetadata(level, message, {
      stackTrace: this.stackTraces,
    });
  }
}
