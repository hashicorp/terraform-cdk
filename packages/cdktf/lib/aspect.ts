// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
//  Originally from aws-cdk v2 because with constructs v10 Aspects where moved to the AWS CDK
//  https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/aspect.ts
import { IConstruct } from "constructs";

const ASPECTS_SYMBOL = Symbol("cdktf-aspects");

/**
 * Represents an Aspect
 */
export interface IAspect {
  /**
   * All aspects can visit an IConstruct
   */
  visit(node: IConstruct): void;
}

/**
 * Aspects can be applied to CDK tree scopes and can operate on the tree before
 * synthesis.
 */
export class Aspects {
  /**
   * Returns the `Aspects` object associated with a construct scope.
   * @param scope The scope for which these aspects will apply.
   */
  public static of(scope: IConstruct): Aspects {
    let aspects = (scope as any)[ASPECTS_SYMBOL];
    if (!aspects) {
      aspects = new Aspects();

      Object.defineProperty(scope, ASPECTS_SYMBOL, {
        value: aspects,
        configurable: false,
        enumerable: false,
      });
    }
    return aspects;
  }

  private readonly _aspects: IAspect[];

  private constructor() {
    this._aspects = [];
  }

  /**
   * Adds an aspect to apply this scope before synthesis.
   * @param aspect The aspect to add.
   */
  public add(aspect: IAspect) {
    this._aspects.push(aspect);
  }

  /**
   * The list of aspects which were directly applied on this scope.
   */
  public get all(): IAspect[] {
    return [...this._aspects];
  }
}
