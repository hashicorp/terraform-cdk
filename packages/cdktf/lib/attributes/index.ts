export * from "./terraform-any-attribute"
export * from "./terraform-any-list-attribute"
export * from "./terraform-any-map-attribute"
export * from "./terraform-any-set-attribute"
export * from "./terraform-attribute"
export * from "./terraform-boolean-attribute"
export * from "./terraform-boolean-list-attribute"
export * from "./terraform-boolean-map-attribute"
export * from "./terraform-boolean-set-attribute"
export * from "./terraform-dynamic-attribute"
export * from "./terraform-list-attribute"
export * from "./terraform-map-attribute"
export * from "./terraform-number-attribute"
export * from "./terraform-number-list-attribute"
export * from "./terraform-number-map-attribute"
export * from "./terraform-number-set-attribute"
export * from "./terraform-object-attribute"
export * from "./terraform-set-attribute"
export * from "./terraform-string-attribute"
export * from "./terraform-string-list-attribute"
export * from "./terraform-string-map-attribute"
export * from "./terraform-string-set-attribute"
