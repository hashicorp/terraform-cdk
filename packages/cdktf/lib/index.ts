// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { validateEnvironment } from "./validateEnvironment";
validateEnvironment();

export * from "./tokens";
export * from "./terraform-stack";
export * from "./terraform-element";
export * from "./terraform-module";
export * from "./terraform-provider";
export * from "./terraform-resource";
export * from "./terraform-data-source";
export * from "./terraform-output";
export * from "./complex-computed-list";
export * from "./resource";
export * from "./testing";
export * as testingMatchers from "./testing/matchers";
export * from "./app";
export * from "./terraform-backend";
export * from "./backends";
export * from "./terraform-remote-state";
export * from "./terraform-local";
export * from "./terraform-variable";
export * from "./terraform-hcl-module";
export * from "./runtime";
export * from "./terraform-dependable";
export * from "./manifest";
export * from "./terraform-asset";
export * from "./synthesize";
export * from "./annotations";
export * from "./aspect";
export * from "./terraform-functions";
export * from "./terraform-operators";
export * from "./tfExpression";
export * from "./terraform-addressable";
export * from "./terraform-iterator";
export * from "./terraform-provisioner";
export * from "./terraform-conditions";
export * from "./terraform-count";

// required for JSII because Fn extends from it
export * from "./functions/terraform-functions.generated";
