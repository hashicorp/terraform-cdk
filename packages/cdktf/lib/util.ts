// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import {
  sourceOrTargetNotAnObject,
  targetNotResolvableWithOverrides,
} from "./errors";
import { TerraformDynamicBlock } from "./terraform-dynamic-block";
import { Tokenization } from "./tokens/token";

export const terraformBinaryName =
  process.env.TERRAFORM_BINARY_NAME || "terraform";

/**
 * Merges `source` into `target`, overriding any existing values.
 * `undefined` will cause a value to be deleted.
 */
export function deepMerge(target: any, ...sources: any[]) {
  if (Tokenization.isResolvable(target) && sources.length > 0) {
    throw targetNotResolvableWithOverrides(target.toString());
  }

  for (const source of sources) {
    if (typeof source !== "object" || typeof target !== "object") {
      throw sourceOrTargetNotAnObject(
        JSON.stringify(source),
        typeof source,
        JSON.stringify(target),
        typeof target
      );
    }

    for (const key of Object.keys(source)) {
      const value = source[key];
      if (typeof value === "object" && value != null && !Array.isArray(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (typeof target[key] !== "object") {
          target[key] = {};
        }

        // if the value is a resolvable we don't want to recurse into it
        if (Tokenization.isResolvable(value)) {
          target[key] = value;
          continue;
        }

        deepMerge(target[key], value);
      } else if (
        typeof value === "object" &&
        value != null &&
        Array.isArray(value)
      ) {
        if (Array.isArray(target[key])) {
          target[key] = [...target[key], ...value];
        } else {
          target[key] = value;
        }
      } else if (value === undefined) {
        delete target[key];
      } else {
        target[key] = value;
      }
    }
  }

  return target;
}

/**
 * Transforms a string to snake case
 */
export function snakeCase(str: string): string {
  if (!str) return "";

  return String(str)
    .replace(/^[^A-Za-z0-9]*|[^A-Za-z0-9]*$/g, "")
    .replace(/([a-z0-9])([A-Z])/g, (_m, a, b) => a + "_" + b.toLowerCase())
    .replace(/[^A-Za-z0-9]+|_+/g, "_")
    .toLowerCase();
}

/**
 * Transforms all keys in a object to snake case
 */
export function keysToSnakeCase(object: any): any {
  if (Tokenization.isResolvable(object)) {
    return object;
  }
  if (Array.isArray(object)) {
    return object.map((e: any) => {
      return typeof e === "object" ? keysToSnakeCase(e) : e;
    });
  }
  const keys = Object.keys(object);
  return keys.reduce((newObject: any, key: string) => {
    if (key === "tags" || key === "environment") {
      newObject[key] = object[key];
      return newObject;
    }

    let value = object[key];
    if (typeof value === "object") {
      value = keysToSnakeCase(value);
    }
    if (Array.isArray(value)) {
      value = value.map((e: any) => {
        return typeof e === "object" ? keysToSnakeCase(e) : e;
      });
    }
    newObject[snakeCase(key)] = value;
    return newObject;
  }, {});
}

/**
 * dynamic attributes are located at a different position than normal block attributes
 * This method detects them and moves them from .attributeName to .dynamic.attributeName
 * It also invokes the .toTerraform() method on the dynamic attribute to get the correct
 * Terraform representation
 */
export function processDynamicAttributesForHcl(attributes: {
  [name: string]: any;
}): {
  [name: string]: any;
} {
  const result: { [name: string]: any; dynamic?: { [name: string]: any } } = {};
  Object.entries(attributes).forEach(([attributeName, value]) => {
    if (TerraformDynamicBlock.isTerraformDynamicBlock(value)) {
      if (!result.dynamic) {
        result.dynamic = {};
      }
      result.dynamic[attributeName] = value.toTerraformDynamicBlockJson();
    } else {
      const recurse =
        value !== null &&
        typeof value === "object" &&
        value.constructor === Object; // only descend into plain objects
      result[attributeName] = recurse
        ? processDynamicAttributesForHcl(value)
        : value;
    }
  });
  return result;
}

/**
 * dynamic attributes are located at a different position than normal block attributes
 * This method detects them and moves them from .attributeName to .dynamic.attributeName
 * It also invokes the .toTerraform() method on the dynamic attribute to get the correct
 * Terraform representation
 */
export function processDynamicAttributes(attributes: { [name: string]: any }): {
  [name: string]: any;
} {
  const result: { [name: string]: any; dynamic?: { [name: string]: any } } = {};
  Object.entries(attributes).forEach(([attributeName, value]) => {
    if (TerraformDynamicBlock.isTerraformDynamicBlock(value)) {
      if (!result.dynamic) {
        result.dynamic = {};
      }
      result.dynamic[attributeName] = value.toTerraformDynamicBlockJson();
    } else {
      const recurse =
        value !== null &&
        typeof value === "object" &&
        value.constructor === Object; // only descend into plain objects
      result[attributeName] = recurse ? processDynamicAttributes(value) : value;
    }
  });
  return result;
}
