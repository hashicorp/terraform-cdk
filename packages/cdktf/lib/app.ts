// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Construct, IConstruct } from "constructs";
import * as fs from "fs";
import { version } from "../package.json";
import { DISABLE_STACK_TRACE_IN_METADATA } from "./annotations";
import { Manifest } from "./manifest";
import { ISynthesisSession } from "./synthesize";
import { TerraformStack } from "./terraform-stack";

const APP_SYMBOL = Symbol.for("cdktf/App");
export const CONTEXT_ENV = "CDKTF_CONTEXT_JSON";
export interface AppConfig {
  /**
   * The directory to output Terraform resources.
   *
   * @default - CDKTF_OUTDIR if defined, otherwise "cdktf.out"
   */
  readonly outdir?: string;
  readonly stackTraces?: boolean;

  /**
   * Additional context values for the application.
   *
   * Context set by the CLI or the `context` key in `cdktf.json` has precedence.
   *
   * Context can be read from any construct using `node.getContext(key)`.
   *
   * @default - no additional context
   */
  readonly context?: { [key: string]: any };

  /**
   * Whether to skip all validations during synthesis of the app
   *
   * @default - false
   */
  readonly skipValidation?: boolean;

  /**
   * Whether to skip backend validation during synthesis of the app
   *
   * @default - false
   */
  readonly skipBackendValidation?: boolean;
}

/**
 * Represents a cdktf application.
 */
export class App extends Construct {
  /**
   * The output directory into which resources will be synthesized.
   */
  public readonly outdir: string;

  /**
   * The stack which will be synthesized. If not set, all stacks will be synthesized.
   */
  public readonly targetStackId: string | undefined;

  public readonly manifest: Manifest;

  /**
   * Whether to skip all validations during synthesis of the app
   */
  public readonly skipValidation: boolean;

  /**
   * Whether to skip backend validation during synthesis of the app
   */
  public readonly skipBackendValidation: boolean;

  /**
   * Defines an app
   * @param config configuration
   */
  constructor(config: AppConfig = {}) {
    super(undefined as any, "");
    Object.defineProperty(this, APP_SYMBOL, { value: true });

    this.outdir = process.env.CDKTF_OUTDIR ?? config.outdir ?? "cdktf.out";
    this.targetStackId = process.env.CDKTF_TARGET_STACK_ID;
    this.skipValidation = config.skipValidation || false;
    this.skipBackendValidation = config.skipBackendValidation || false;

    this.loadContext(config.context);

    const node = this.node;
    if (config.stackTraces === false) {
      node.setContext(DISABLE_STACK_TRACE_IN_METADATA, true);
    }

    node.setContext("cdktfVersion", version);

    if (!fs.existsSync(this.outdir)) {
      fs.mkdirSync(this.outdir);
    }
    this.manifest = new Manifest(version, this.outdir);
  }

  public static isApp(x: any): x is App {
    return x !== null && typeof x === "object" && APP_SYMBOL in x;
  }

  public static of(construct: IConstruct): App {
    return _lookup(construct);

    // eslint-disable-next-line jsdoc/require-jsdoc
    function _lookup(c: IConstruct): App {
      if (App.isApp(c)) {
        return c;
      }

      const node = c.node;

      if (!node.scope) {
        throw new Error(
          `No app could be identified for the construct at path '${construct.node.path}'`
        );
      }

      return _lookup(node.scope);
    }
  }

  /**
   * Synthesizes all resources to the output directory
   */
  public synth(): void {
    const session: ISynthesisSession = {
      outdir: this.outdir,
      skipValidation: this.skipValidation,
      manifest: this.manifest,
    };

    const stacks = this.node
      .findAll()
      .filter<TerraformStack>(TerraformStack.isStack);

    stacks.forEach((stack) => stack.prepareStack());
    stacks.forEach((stack) => stack.synthesizer.synthesize(session));

    if (!this.skipValidation) {
      const validations = this.node.validate();
      if (validations.length) {
        const errorList = validations.join("\n  ");
        throw new Error(
          `App level validation failed with the following errors:\n  ${errorList}`
        );
      }
    }

    this.manifest.writeToFile();
  }

  /**
   * Creates a reference from one stack to another, invoked on prepareStack since it creates extra resources
   */
  public crossStackReference(
    fromStack: TerraformStack,
    toStack: TerraformStack,
    identifier: string
  ): string {
    toStack.addDependency(fromStack);
    const outputId =
      fromStack.registerOutgoingCrossStackReference(
        identifier
      ).friendlyUniqueId;

    const remoteState = toStack.registerIncomingCrossStackReference(fromStack);

    return remoteState.getString(outputId);
  }

  private loadContext(defaults: { [key: string]: string } = {}) {
    const node = this.node;

    // prime with defaults passed through constructor
    for (const [k, v] of Object.entries(defaults)) {
      node.setContext(k, v);
    }

    // read from environment
    const contextJson = process.env[CONTEXT_ENV];
    const contextFromEnvironment = contextJson ? JSON.parse(contextJson) : {};

    for (const [k, v] of Object.entries(contextFromEnvironment)) {
      node.setContext(k, v);
    }
  }
}
