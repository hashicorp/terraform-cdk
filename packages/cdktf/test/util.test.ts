// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { deepMerge } from "../lib/util";

test("merge arrays", () => {
  const target = {
    foo: [1, 2],
  };

  const source = {
    foo: [3, 4],
  };

  expect(JSON.stringify(deepMerge(target, source))).toMatchSnapshot();
});

test("merge objects", () => {
  const target = {
    foo: {
      bar: 2,
    },
  };

  const source = {
    foo: {
      baz: 3,
    },
  };

  expect(JSON.stringify(deepMerge(target, source))).toMatchSnapshot();
});
