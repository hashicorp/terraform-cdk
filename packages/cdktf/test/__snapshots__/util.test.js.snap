// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge arrays 1`] = `"{"foo":[1,2,3,4]}"`;

exports[`merge objects 1`] = `"{"foo":{"bar":2,"baz":3}}"`;
